### PR TITLE
Fixing issues with lost references when serialising builders.

### PR DESCRIPTION
**Pull request by keriksson-rosenqvist**
Porting changes from develop for this fix.

### CODE REVIEW

File: src/QAT/qat/purr/compiler/devices.py [hunk 1]
```diff
@@ -234,7 +234,7 @@ def _log_ref(self, obj):
         is_new = super()._log_ref(obj)
         # We don't want complex' or enums to be referenced. Just create a new one on
         # each instance.
-        if isinstance(obj, (complex, Enum)):
+        if isinstance(obj, (complex, Enum, np.number)):
```
Comment (jfriel-oqc) [inline src/QAT/qat/purr/compiler/devices.py:237]: what does numpy number capture here that complex doesn't? Just curious

File: src/QAT/qat/purr/compiler/instructions.py [hunk 2]
```diff
@@ -396,7 +392,10 @@ def __init__(self, acquire: Acquire, process, axes=None, args=None):
             axes = [axes]
 
         self.process: PostProcessType = process
-        self.args: List = args or []
+
+        # TODO: Should find out why ags can't reference model objects in builder serialization.
```
Comment (jfriel-oqc) [inline src/QAT/qat/purr/compiler/instructions.py:396]: can we put an issue in for this as a bug, cheers

File: src/tests/test_qasm.py [hunk 3]
```diff
@@ -851,6 +852,21 @@ def test_execute_different_qat_input_types(self):
...
+        hardware = get_default_echo_hardware(8)
+        config = CompilerConfig()
+
+        frontend = QASMFrontend()
+        builder, metrics = frontend.parse(qasm_string, hardware, config)
+
+        serialized_builder = builder.serialize()
+        builder = QuantumInstructionBuilder.deserialize(serialized_builder)
+
+        results = frontend.execute(builder, hardware, config)
```
Comment (jfriel-oqc) [inline src/tests/test_qasm.py:866]: the builder has a model/hardware on it. I think both cases should be tested here, that you use the original hardward and you also use the rehydrated hardware from the builder.

**Review by jfriel-oqc — APPROVED** [review]
approving up to an additional test requested for sanity purposes

**Review by hamidelmaazouz — APPROVED** [review]

**Comment by hamidelmaazouz** [comment]
could you also close the other PR at https://github.com/oqc-community/qat/pull/78 ?

cc/ @chemix-lunacy

File: src/QAT/qat/purr/compiler/devices.py (hunk 1 above)
Comment (keriksson-rosenqvist) [reply src/QAT/qat/purr/compiler/devices.py:237]: Numpy wrapped numbers, e.g. `np.float64`

File: src/QAT/qat/purr/compiler/instructions.py (hunk 2 above)
Comment (keriksson-rosenqvist) [reply src/QAT/qat/purr/compiler/instructions.py:396]: Done. https://github.com/oqc-community/qat/issues/86

**Pull request merged**